File: agent/custom_openai_client.py
import requests
import json
import os
from typing import Dict, List, Any, Optional


class ChatCompletion:
    def __init__(self, response_data: Dict):
        self.choices = [Choice(choice) for choice in response_data.get("choices", [])]
        self.id = response_data.get("id")
        self.object = response_data.get("object")
        self.created = response_data.get("created")
        self.model = response_data.get("model")
        self.usage = response_data.get("usage")


class Choice:
    def __init__(self, choice_data: Dict):
        self.message = Message(choice_data.get("message", {}))
        self.index = choice_data.get("index")
        self.finish_reason = choice_data.get("finish_reason")


class Message:
    def __init__(self, message_data: Dict):
        self.content = message_data.get("content")
        self.role = message_data.get("role")


class ChatCompletions:
    def __init__(self, api_client):
        self.api_client = api_client

    def create(
        self,
        model: str,
        messages: List[Dict],
        max_tokens: int = 1000,
        temperature: float = 0.1,
        **kwargs,
    ) -> ChatCompletion:
        """Create a chat completion using the custom API"""
        return self.api_client._make_chat_request(
            model=model,
            messages=messages,
            max_tokens=max_tokens,
            temperature=temperature,
            **kwargs,
        )


class Chat:
    def __init__(self, api_client):
        self.completions = ChatCompletions(api_client)


class CustomOpenAIClient:
    def __init__(
        self, api_key: str = None, base_url: str = "https://api.omnia.reainternal.net"
    ):
        self.api_key = api_key or os.getenv("OPENAI_API_KEY")
        self.base_url = base_url.rstrip("/")
        self.chat = Chat(self)

        if not self.api_key:
            raise ValueError(
                "API key is required. Set OPENAI_API_KEY environment variable or pass api_key parameter."
            )

    def _get_headers(self) -> Dict[str, str]:
        """Get headers for API requests"""
        return {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json",
            "User-Agent": "CustomOpenAIClient/1.0",
        }

    def _make_chat_request(
        self,
        model: str,
        messages: List[Dict],
        max_tokens: int = 1000,
        temperature: float = 0.1,
        **kwargs,
    ) -> ChatCompletion:
        """Make a chat completion request to the API"""

        url = f"{self.base_url}/v1/chat/completions"

        payload = {
            "model": model,
            "messages": messages,
            "max_tokens": max_tokens,
            "temperature": temperature,
            **kwargs,
        }

        headers = self._get_headers()

        try:
            response = requests.post(url, headers=headers, json=payload, timeout=60)
            response.raise_for_status()

            response_data = response.json()
            return ChatCompletion(response_data)

        except requests.exceptions.RequestException as e:
            if hasattr(e, "response") and e.response is not None:
                try:
                    error_data = e.response.json()
                    error_message = error_data.get("error", {}).get("message", str(e))
                except:
                    error_message = f"HTTP {e.response.status_code}: {e.response.text}"
            else:
                error_message = str(e)

            raise Exception(f"API request failed: {error_message}")
        except json.JSONDecodeError as e:
            raise Exception(f"Failed to parse API response: {str(e)}")
        except Exception as e:
            raise Exception(f"Unexpected error: {str(e)}")


# Alias for compatibility
OpenAI = CustomOpenAIClient

File: agent/custom_openai_client.js
const https = require("https");
const http = require("http");
const { URL } = require("url");

class ChatCompletion {
  constructor(responseData) {
    this.choices = (responseData.choices || []).map(
      (choice) => new Choice(choice)
    );
    this.id = responseData.id;
    this.object = responseData.object;
    this.created = responseData.created;
    this.model = responseData.model;
    this.usage = responseData.usage;
  }
}

class Choice {
  constructor(choiceData) {
    this.message = new Message(choiceData.message || {});
    this.index = choiceData.index;
    this.finish_reason = choiceData.finish_reason;
  }
}

class Message {
  constructor(messageData) {
    this.content = messageData.content;
    this.role = messageData.role;
  }
}

class ChatCompletions {
  constructor(apiClient) {
    this.apiClient = apiClient;
  }

  async create({
    model,
    messages,
    max_tokens = 1000,
    temperature = 0.1,
    ...kwargs
  }) {
    return await this.apiClient._makeChatRequest({
      model,
      messages,
      max_tokens,
      temperature,
      ...kwargs,
    });
  }
}

class Chat {
  constructor(apiClient) {
    this.completions = new ChatCompletions(apiClient);
  }
}

class CustomOpenAIClient {
  constructor({
    apiKey = null,
    baseURL = "https://api.omnia.reainternal.net",
  } = {}) {
    this.apiKey = apiKey || process.env.OPENAI_API_KEY;
    this.baseURL = baseURL.replace(/\/$/, "");
    this.chat = new Chat(this);

    if (!this.apiKey) {
      throw new Error(
        "API key is required. Set OPENAI_API_KEY environment variable or pass apiKey parameter."
      );
    }
  }

  _getHeaders() {
    return {
      Authorization: `Bearer ${this.apiKey}`,
      "Content-Type": "application/json",
      "User-Agent": "CustomOpenAIClient/1.0",
    };
  }

  async _makeChatRequest({
    model,
    messages,
    max_tokens = 1000,
    temperature = 0.1,
    ...kwargs
  }) {
    const url = `${this.baseURL}/v1/chat/completions`;

    const payload = {
      model,
      messages,
      max_tokens,
      temperature,
      ...kwargs,
    };

    const headers = this._getHeaders();

    return new Promise((resolve, reject) => {
      const urlObj = new URL(url);
      const protocol = urlObj.protocol === "https:" ? https : http;

      const options = {
        hostname: urlObj.hostname,
        port: urlObj.port,
        path: urlObj.pathname + urlObj.search,
        method: "POST",
        headers: {
          ...headers,
          "Content-Length": Buffer.byteLength(JSON.stringify(payload)),
        },
        timeout: 60000,
      };

      const req = protocol.request(options, (res) => {
        let data = "";

        res.on("data", (chunk) => {
          data += chunk;
        });

        res.on("end", () => {
          try {
            if (res.statusCode >= 200 && res.statusCode < 300) {
              const responseData = JSON.parse(data);
              resolve(new ChatCompletion(responseData));
            } else {
              let errorMessage;
              try {
                const errorData = JSON.parse(data);
                errorMessage =
                  errorData.error?.message || `HTTP ${res.statusCode}: ${data}`;
              } catch {
                errorMessage = `HTTP ${res.statusCode}: ${data}`;
              }
              reject(new Error(`API request failed: ${errorMessage}`));
            }
          } catch (parseError) {
            reject(
              new Error(`Failed to parse API response: ${parseError.message}`)
            );
          }
        });
      });

      req.on("error", (error) => {
        reject(new Error(`Request error: ${error.message}`));
      });

      req.on("timeout", () => {
        req.abort();
        reject(new Error("Request timeout"));
      });

      req.write(JSON.stringify(payload));
      req.end();
    });
  }
}

// Export for compatibility
class OpenAI extends CustomOpenAIClient {}

module.exports = { OpenAI, CustomOpenAIClient };
